In [ ]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download(
    "balaka18/email-spam-classification-dataset-csv"
)

print("Path to dataset files:", path)

100%|██████████| 1.66M/1.66M [00:00<00:00, 111MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/balaka18/email-spam-classification-dataset-csv/versions/1


In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sn

from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import (
    accuracy_score,
    confusion_matrix,
    classification_report
)

In [ ]:
import os

file_path = os.path.join(path, "emails.csv")

df = pd.read_csv(file_path)

print("Dataset shape:", df.shape)
print(df.head())

Dataset shape: (5172, 3002)
  Email No.  the  to  ect  and  for  of    a  you  hou  ...  connevey  jay  \
0   Email 1    0   0    1    0    0   0    2    0    0  ...         0    0   
1   Email 2    8  13   24    6    6   2  102    1   27  ...         0    0   
2   Email 3    0   0    1    0    0   0    8    0    0  ...         0    0   
3   Email 4    0   5   22    0    5   1   51    2   10  ...         0    0   
4   Email 5    7   6   17    1    5   2   57    0    9  ...         0    0   

   valued  lay  infrastructure  military  allowing  ff  dry  Prediction  
0       0    0               0         0         0   0    0           0  
1       0    0               0         0         0   1    0           0  
2       0    0               0         0         0   0    0           0  
3       0    0               0         0         0   0    0           0  
4       0    0               0         0         0   1    0           0  

[5 rows x 3002 columns]


In [ ]:
print("Columns:", len(df.columns))
print("\nDataset information:")
print(df.info())

print("\nClass distribution:")
print(df.iloc[:, -1].value_counts())

Columns: 3002

Dataset information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5172 entries, 0 to 5171
Columns: 3002 entries, Email No. to Prediction
dtypes: int64(3001), object(1)
memory usage: 118.5+ MB
None

Class distribution:
Prediction
0    3672
1    1500
Name: count, dtype: int64


In [ ]:
X = df.iloc[:, 1:-1]
y = df.iloc[:, -1]

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (5172, 3000)
y shape: (5172,)


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Training data:", X_train.shape)
print("Testing data:", X_test.shape)

Training data: (4137, 3000)
Testing data: (1035, 3000)


In [ ]:
model = SVC(kernel='linear')

In [ ]:
model.fit(X_train, y_train)

print("SVM training completed!")

SVM training completed!


In [ ]:
y_pred = model.predict(X_test)

print("Predicted values:")
print(y_pred)

In [ ]:
accuracy = accuracy_score(y_test, y_pred)

print("Accuracy Score:", accuracy)
print("Accuracy Percentage:", accuracy * 100, "%")

In [ ]:
cm = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(cm)
plt.figure(figsize=(6, 4))

sn.heatmap(
    cm,
    annot=True,
    fmt='d',
    xticklabels=['Not Spam', 'Spam'],
    yticklabels=['Not Spam', 'Spam']
)

plt.xlabel("Predicted")
plt.ylabel("Actual")
plt.title("SVM Confusion Matrix")

plt.show()

In [ ]:
print("Classification Report:")
print(
    classification_report(
        y_test,
        y_pred,
        target_names=['Not Spam', 'Spam']
    )
)

In [ ]:
sample = X_test.iloc[[0]]

prediction = model.predict(sample)[0]

if prediction == 1:
    print("Prediction: SPAM")
else:
    print("Prediction: NOT SPAM")

### Summary of What We Did and How We Did It

#### What We Did
We built and evaluated a Machine Learning model to classify emails as **Spam** (1) or **Not Spam** (0).

#### How We Did It
1. **Data Acquisition:** Downloaded the "Email Spam Classification" dataset from Kaggle using `kagglehub`.
2. **Data Loading:** Loaded the `emails.csv` file containing 5,172 emails and 3,002 columns (word frequencies and labels).
3. **Preprocessing:** Split the dataset into features (`X` containing 3,000 word frequencies) and labels (`y` containing the target prediction).
4. **Data Splitting:** Divided the data into training (80%) and testing (20%) sets using stratified sampling to maintain class balance.
5. **Model Training:** Trained a Support Vector Classifier (SVC) with a `linear` kernel on the training set.
6. **Evaluation:** Evaluated the model on the testing set using Accuracy, a Confusion Matrix heatmap, and a Classification Report (Precision, Recall, F1-Score).